source/common/router/router.cc

```diff
@@ -487,7 +488,24 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Fetch a connection pool for the upstream cluster.
-  Http::ConnectionPool::Instance* conn_pool = getConnPool();
+  const auto http_protocol_options = cluster_->httpProtocolOptions();
+
+  if (http_protocol_options.has_value() && http_protocol_options.value().auto_sni()) {
+    const auto host_str = headers.Host()->value().getStringView();
+    const auto parsed_authority = Http::Utility::parseAuthority(host_str.data());
+
+    if (!parsed_authority.is_ip_address_) {
+      // Update filter state with the host/authority to use for setting SNI in the transport socket
+      // options. This is referenced during the getConnPool() call below.
+      callbacks_->streamInfo().filterState().setData(
+          Network::UpstreamServerName::key(),
+          std::make_unique<Network::UpstreamServerName>(host_str),
```

Comment:
Does this also force SAN verification as we do in:
https://github.com/envoyproxy/envoy/blob/f08b17516026882a0964a225c0135dc996b964af/source/extensions/clusters/dynamic_forward_proxy/cluster.cc#L102-L107

I think it should and optimally we should have an integration test that covers this case. cc @lizan

Reply:
Not yet, we should do that in a follow up PR of this to support complete SNI/SAN combination, and refactor dynamic forward cluster to leverage this in router.